src/cli/cli_udp.cpp

```diff
@@ -172,17 +157,65 @@ otError UdpExample::ProcessSend(int argc, char *argv[])
         messageInfo.mPeerPort = static_cast<uint16_t>(value);
     }
 
+    if (argc == 2 || argc == 4)
+    {
+        int typePos = curArg++;
+
+        if (strcmp(argv[typePos], "-s") == 0)
+        {
+            unsigned long value;
+            payloadType = kTypeAutoSize;
+            SuccessOrExit(error = Interpreter::ParseUnsignedLong(argv[curArg], value));
+            payloadLength = static_cast<uint16_t>(value);
+        }
+        else if (strcmp(argv[typePos], "-h") == 0)
+        {
+            // Assume one byte would be expressed by two chars in HexString.
+            // Here HexString payload should have even size.
+            payloadLength = static_cast<uint16_t>(strlen(argv[curArg]));
+            VerifyOrExit((payloadLength & 0x0001) == 0, error = OT_ERROR_INVALID_ARGS);
+            payloadType = kTypeHexString;
+        }
+        else if (strcmp(argv[typePos], "-t") == 0)
+        {
+            payloadType = kTypeText;
+        }
+    }
+
     message = otUdpNewMessage(mInterpreter.mInstance, NULL);
     VerifyOrExit(message != NULL, error = OT_ERROR_NO_BUFS);
 
-    if (autoGenPayload)
-    {
-        error = WriteCharToBuffer(message, static_cast<uint16_t>(autoGenPayloadLength));
-    }
-    else
+    switch (payloadType)
     {
+    case kTypeText:
         error = otMessageAppend(message, argv[curArg], static_cast<uint16_t>(strlen(argv[curArg])));
+        break;
+    case kTypeAutoSize:
+        error = WriteCharToBuffer(message, payloadLength);
+        break;
+    case kTypeHexString:
+    {
+        // Generally the rx buffer of uart would be 128.
+        // The length of the hexString in bytes would always be less than 128/2=64.
+        // Here uses one array of 70 bytes to store the payload.
+        uint8_t     buf[70];
+        uint8_t     bufLen  = 0;
+        uint8_t     index   = 0;
+        const char *payload = argv[curArg];
+
+        while (index < payloadLength && bufLen < sizeof(buf))
+        {
+            buf[bufLen] = ((*(payload + index)) - '0') << 4 | ((*(payload + index + 1)) - '0');
```

Comment:
1. This doesn't cover `a-f` and `A-F`;
2. Should use `Interpreter::Hex2Bin()` instead;

Reply:
thanks, updated.